File: 0074_Search_a_2D_Matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        def searchRow(matrix: List[List[int]], target: int) -> List[int]:
            N = len(matrix)
            if N == 1:
                return matrix[0]
            
            midi = N // 2
            if matrix[midi][0] > target:
                return searchRow(matrix[:midi], target)
            else:
                return searchRow(matrix[midi:], target)
            
        def searchCol(ls: List[int], target: int) -> bool:
            print(ls)
            M = len(ls)
            if M == 1 and ls[0] == target:
                return True
            elif M == 1: 
                return False
            midi = M // 2
            if ls[midi] > target:
                return searchCol(ls[:midi], target)
            else:
                return searchCol(ls[midi:], target)
        if len(matrix) == 0 or len(matrix[0]) == 0:
            return False
        
        return searchCol(searchRow(matrix, target), target)

File: 0074_Search_a_2D_Matrix.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        if (matrix.length == 0) return false;
        int N = matrix[0].length;
        int lo = 0, hi = matrix.length * N;
        while (lo < hi) {
            int mid = (lo + hi) / 2;
            int x = matrix[mid/N][mid%N];
            if (x < target) {
                lo = mid + 1;
            } else if (x > target) {
                hi = mid;
            } else {
                return true;
            }
        }
        return false;
    }
}